kube-runtime/src/controller/mod.rs

```diff
@@ -230,22 +234,25 @@ where
     QueueStream::Error: std::error::Error + 'static,
 {
     let (scheduler_shutdown_tx, scheduler_shutdown_rx) = channel::oneshot::channel();
-    let err_context = context.clone();
-    let (scheduler_tx, scheduler_rx) = channel::mpsc::unbounded::<ScheduleRequest<ReconcileRequest<K>>>();
+    let (scheduler_tx, scheduler_rx) =
+        channel::mpsc::channel::<ScheduleRequest<ReconcileRequest<K>>>(APPLIER_REQUEUE_BUF_SIZE);
```

Comment:
some points that i wanted to lift regarding the unbounded channels here because i'm not sure about the full reasoning here. it feels partly defensible to have unbounded queues:

- there's __generally__ a limit to how many items people put inside clusters because of how impractical it is to have a large-scale one-to-many relations with common items that have realistic cluster bounds (pods generally do not grow forever, people often make new clusters at some point after 10k or higher)
- large numbers of objects make global reconcilers an O(n^2) problem. constant requeuing, and retriggering in such cases are also likely to waste of resources (node affecting controllers in controller manager are a huge IO hogs in particular)
- if we are in such a large-scale case where we have 10k+ items in our reflector cache, and we want to reconcile all of them, we first need enough memory to house 10k+ of these specs, and an unbounded queue would at most double resources

but on the other hand. if i understand this queue correctly, it also serves as a limiter of the amount of parallelism in a controller? in that case, limiting it actually makes a lot of sense, because 10k objects being reconciled at once might DoS a third party service.

Reply:
Not quite, there are actually four different "queues" going on in `applier`:

1. The executor itself, this has no hard limit but is de-duped (so there will only be one task running at any one moment per K8s object)
2. The "pending reconciliations" queue (for objects that are already reconciling but should be retried as soon as their current job is done) queue, this is also unbounded but deduped
3. The scheduler (for objects that are scheduled to be reconciled again at some point in the future because they requested it via `Action`), this is also unbounded but deduped
4. The "pending scheduling requests" queue, for objects that haven't been organized into one of the previous three queues yet

This PR is only concerned with queue 4, where we have no practical way to implement deduping.

Reply:
So if I understand it right:

- the pending reconciliations queue is `pending` in `Scheduler` (i.e. 2)
- the internal `DelayQueue` on `Scheduler` is 3
- 4 is effectively the queuestream; the inputs 

and the executor is going to work at its regular pace. ..so this means it is actually possible to reconcile 1000s of reconciles at the same time on re-lists currently? :grimacing:. Is that something that is viable to bound at some point, somehow? In the `Runner`?

Reply:
> So if I understand it right:

Yes.

> and the executor is going to work at its regular pace. ..so this means it is actually possible to reconcile 1000s of reconciles at the same time on re-lists currently? :grimacing:

Yes.

> Is that something that is viable to bound at some point, somehow? In the Runner?

Well, we could add additional constraints for when we actually start running a pending reconciliation. That's not implemented at the moment, on the assumption that you could "just" use a [semaphore](https://docs.rs/tokio/latest/tokio/sync/struct.Semaphore.html) in your reconciler function.

Reply:
Doing the semaphoring in the executor (which has other benefits, like not having to allocate the task before it actually has a semaphore permit) shouldn't be too difficult either, the main problem there would be that `applier`'s configuration is already getting pretty unwieldy as it is.

Reply:
Maybe it makes sense to refactor a lot of appliers configuration into some kind of `ApplierParams` struct that can be heavily defaulted.

Reply:
Probably, yeah..